From 0ee9f6cc8bc156693f090b893a380c06ad9d88d2 Mon Sep 17 00:00:00 2001
From: Rahul Bhattacharjee <quic_rbhattac@quicinc.com>
Date: Tue, 13 Dec 2022 15:53:02 +0530
Subject: [PATCH] ath12k: Monitor mode bring up in SLO

Currently all the radios are sharing the same chandef, so during monitor
start wrong values are getting passed to FW, due to which we are seeing
FW assert.

Fixed by adding a check, when ar is not compatible with the chandef,
it will stop the process.

In addition, cfg80211 was not allowing to change the channels before
bringing up the interface, it needs at least one interface up to
change the channels.

Hence, removing the check for number of running interfaces.

Signed-off-by: Rahul Bhattacharjee <quic_rbhattac@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 4 ++++
 net/wireless/core.h                   | 3 +--
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index ef1218a..1b4bd09 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1233,6 +1233,10 @@ static int ath12k_mac_monitor_start(struct ath12k *ar)
 	if (!chandef)
 		return 0;
 
+	/* TODO 5G low high split changes */
+	if (!ar->mac.sbands[chandef->chan->band].channels)
+		return -EINVAL;
+
 	ret = ath12k_mac_monitor_vdev_start(ar, ar->monitor_vdev_id, chandef);
 	if (ret) {
 		ath12k_warn(ar->ab, "failed to start monitor vdev: %d\n", ret);
diff --git a/net/wireless/core.h b/net/wireless/core.h
index 55c535f..62101da 100644
--- a/net/wireless/core.h
+++ b/net/wireless/core.h
@@ -239,8 +239,7 @@ static inline bool cfg80211_has_monitors_only(struct cfg80211_registered_device
 {
 	lockdep_assert_held(&rdev->wiphy.mtx);
 
-	return rdev->num_running_ifaces == rdev->num_running_monitor_ifaces &&
-	       rdev->num_running_ifaces > 0;
+	return rdev->num_running_ifaces == rdev->num_running_monitor_ifaces;
 }
 
 enum cfg80211_event_type {
-- 
2.38.0

